src/lib.rs

```diff
@@ -22,52 +56,76 @@ fn calculate_hash<T: Hash>(t: &T) -> u64 {
-    }
+    };
 
-    let local_name = format_ident!("local_{}", calculate_hash(&ty));
-    inner_types.push(quote! {
-        let #local_name = <#ty>::add_to_database(database)?;
-    });
+    set.push(ty);
 }
 
 /// Implements AddToTypeDatabase for the type. If the type is a structure, it will
 /// create a new type for the structure with the same name and all its inner fields.
 #[proc_macro_derive(AddToTypeDatabase, attributes(field))]
 pub fn derive_add_to_database(input: TokenStream) -> TokenStream {
-    let input: DeriveInput = parse_macro_input!(input);
-    let name = &input.ident;
+    let receiver = match Receiver::from_derive_input(&parse_macro_input!(input)) {
+        Ok(v) => v,
+        Err(e) => return e.write_errors().into(),
+    };
 
-    let mut inner_types = vec![];
+    // A list of all the types that are being registered with the database.
+    // This list can deliberately include duplicates; these will end up
+    // being shadowed variables. This duplication is needed because each
+    // occurrence of a type will have a distinct span where the type was
+    // encountered in the input, and deduping the types would only result
+    // in the first occurrence of the type getting a compile error.
+    let mut types = vec![];
     let mut fields = vec![];
-    match input.data {
-        syn::Data::Struct(s) => {
-            for field in s.fields {
-                let field_name = field.ident.expect("Field has no name").to_string();
-                let ty = field.ty;
-                add_type(&ty, &mut inner_types);
-                let local_name = format_ident!("local_{}", calculate_hash(&ty));
-                fields.push(quote! {(#field_name, #local_name)});
-            }
-        }
-        _ => panic!("Not a structure."),
+    for Field { ident, ty, .. } in receiver
+        .data
+        .as_ref()
+        .take_struct()
+        .expect("Only named structs supported")
+    {
+        add_type(ty, &mut types);
+
+        let field_name = LitStr::new(
+            &ident.as_ref().expect("Field has no name").to_string(),
+            ident.span(),
+        );
+
+        let local_name = local_ident(&ty);
+
+        fields.push(quote! {(#field_name, #local_name)});
     }
 
+    let name = &receiver.ident;
+    let crate_root = &receiver.crate_root;
+    let db_param = Ident::new("database", Span::call_site());
+    let registrations = types.iter().map(|ty| TypeRegistration {
+        ty,
+        database: &db_param,
+    });
+
     let gen = quote! {
-        impl bpf_script::types::AddToTypeDatabase for #name {
-            fn add_to_database(database: &mut bpf_script::types::TypeDatabase) -> bpf_script::error::Result<usize> {
-                #(#inner_types)*
-                let struct_fields = [#(#fields),*].to_vec();
-                database.add_struct_by_ids(Some(stringify!(#name)), struct_fields.as_slice())
+        #[automatically_derived]
+        impl #crate_root::types::AddToTypeDatabase for #name {
```

Comment:
Right now, this blindly assumes that it won't be invoked on a struct that has any generics, and it'll fail to compile if there are type, lifetime, or const generics present. My reading of what the macro does suggests it probably would not behave well with type parameters, so it might be worth a follow-up PR that adds `generics` to `Receiver`, and requires they be empty before it emits anything.

Reply:
Thanks so much for this! This derive macro is mostly used with simple structures/types so it's all good. Working with generics would be a bonus.

Given my available free time, I spread myself very thin with this set of BPF-related projects so there's a lot of questionable / rushed stuff around. I really appreciate you taking the time to make it just that much better!